packages/react-router/index.js

```diff
@@ -188,6 +188,7 @@ export function Router({ children = null, history, timeout = 2000 }) {
   let [location, setLocation] = React.useState(history.location);
   let [startTransition, pending] = useTransition({ timeoutMs: timeout });
   let listeningRef = React.useRef(false);
+  let unlisten = null;
```

Comment:
Just curious, Is `unlisten` going to be `null` on the next render, since line 201 won't get called again?

We probably need to use `useRef` to retain the unsubscribe function or assign to it to ` listeningRef.current`

Reply:
`useEffect` here is a closure. It maintains a reference to `unlisten` when it is executed which is when it is mounted (and it's the only time that this effect is executed before returning on unmounted).

For example, I logged at following parts:
![image](https://user-images.githubusercontent.com/237753/76767210-9f4ee500-676f-11ea-9ccb-cdd193ccdc99.png)

Then, ran with several re-rendering before unmounting.
![image](https://user-images.githubusercontent.com/237753/76767355-d8875500-676f-11ea-8bd9-4babd514f354.png)

Reply:
Actually, we may not need Ref at all for the purpose of subscription/un-subscription.

```
useEffect(() => {
  const unlisten = history.listen(({ location }) => {
    startTransition(() => {
      setLocation(location);
    });
  });
  return () => {
    if (unlisten != null) {
      unlisten();
    }
  };
}, []);
```

Reply:
By the way, your suggestion to attach `unlisten` to `listeningRef.current` would make the code cleaner too. Let me refactor.